pyogrio/_io.pyx

```diff
@@ -1384,19 +1407,27 @@ def ogr_open_arrow(
         # make sure layer is read from beginning
         OGR_L_ResetReading(ogr_layer)
 
-        if not OGR_L_GetArrowStream(ogr_layer, &stream, options):
-            raise RuntimeError("Failed to open ArrowArrayStream from Layer")
+        # allocate the stream struct and wrap in capsule to ensure clean-up on error
+        if return_capsule:
+            capsule = alloc_c_stream(&stream)
+        else:
+            stream = <ArrowArrayStream*> malloc(sizeof(ArrowArrayStream))
 
-        stream_ptr = <uintptr_t> &stream
+        if not OGR_L_GetArrowStream(ogr_layer, stream, options):
+            free(stream)
+            raise RuntimeError("Failed to open ArrowArrayStream from Layer")
 
         if skip_features:
             # only supported for GDAL >= 3.8.0; have to do this after getting
             # the Arrow stream
             OGR_L_SetNextByIndex(ogr_layer, skip_features)
 
-        # stream has to be consumed before the Dataset is closed
-        import pyarrow as pa
-        reader = pa.RecordBatchStreamReader._import_from_c(stream_ptr)
+        if return_capsule:
+            reader = capsule
```

Comment:
To simplify the return objects, can we always return the bare capsule from here? 

Then in the Python code if the user wants a pyarrow object, you could call `pa.RecordBatchStreamReader._import_from_c(stream_ptr)`? Or with pyarrow 15 pass to your new `from_stream` constructor?

Reply:
Yes, that would be nicer, but that requires a minimum of pyarrow 14.0 to be able to use `RecordBatchReader._import_from_c_capsule`, because in python space we can't easily get the pointer out of the capsule. So therefore left that code path as is for now

Reply:
Although for testing purposes, we have been using `ctypes.pythonapi.PyCapsule_..` methods, so maybe that way we can actually do that in Python. The other problem is that we have to somehow keep the capsule alive (pyarrow doesn't allow to attach it to the RecordBatchReader), and just having it as a variable in `ogr_open_arrow` before the `yield` didn't seem to be sufficient (although for the `ogr_dataset` it is, not sure if cython handles python objects differently ..)